New Component: Button with dropdown

kstinson14 opened the pull request:
On the field apps web team, we've been trying to integrate calcite components into our apps.  We also wanted to try getting our feet wet building some of these components.  So I tried my hand at making a simple button with an overflow menu.  This seemed like a simple and widely-usable enough component that I thought it could be a good fit here.
![Screen Shot 2020-03-13 at 12 38 11 PM](https://user-images.githubusercontent.com/17748358/76641182-8fdd5b00-6527-11ea-83f5-1cbef73d7847.png)

Essentially, it's a button squished together with a dropdown component, and styled so that they maintain the same heights.  We're using it for a control that has a primary action of "Save" but also secondary actions in the dropdown menu like "Discard Changes" or "Clear all".

I'm sure it could be used in several places throughout the platform.  I didn't do a large search or anything, but there is a similar control in the home app for example that could use this component or something similar: 
![image (3)](https://user-images.githubusercontent.com/17748358/76640766-e8602880-6526-11ea-890b-3c9951e28c81.png)

The component has examples in the dummy app, storybook examples, tests, rtl support, theming (currently supports colors/themes that `calcite-button` does), and accessibility support as well.

There were a few tweaks to other components I needed to make to get this to work as intended.
- One was adding a mutation observer to `calcite-button` to observe the button's text content.  In our usage of this control, we're having the primary button have text and an icon on larger screens, and only an icon on smaller screens.  In order to make the button's margins, etc. update correctly as the screen resizes and the text is removed, it was necessary to make `hasText` a dynamic property.
- I needed to tweak the rtl styling for `calcite-button`.  I think some of the rtl selectors in there were missed in the recent PR that switched them over to `host-context`, so those should be fixed now.
- I had to add a `stopPropagation` in addition to the `preventDefault` in `calcite-dropdown`'s `openDropdown` method.  The `preventDefault` didn't seem to be doing the trick in preventing a second click event, so the dropdown kept opening but immediately closing again.

macandcheese commented:
Wow, awesome @kstinson14 ! This is definitely a good element to standardize. 

My initial thought is, we are doing a lot of prop-passing here, would it be better to maybe use slots that let folks set their desired props directly on the individual components:

```
 render() {
    const dir = getElementDir(this.el);
    return (
      <Host dir={dir}>
        <div>
          <slot name="button"></slot>
          <div class="button-dropdown__divider-container">
            <div class="button-dropdown__divider" />
          </div>
          <slot name="dropdown"></slot>
        </div>
      </Host>
    );
  }
```
Then as a consumer:

```
 <calcite-button-with-dropdown>
      <calcite-button slot="button">Here</calcite-button>
      <calcite-dropdown alignment="end" slot="dropdown">
        <calcite-button slot="dropdown-trigger" icon="caretDown"></calcite-button>
        <calcite-dropdown-item>Option 2</calcite-dropdown-item>
        <calcite-dropdown-item>Option 3</calcite-dropdown-item>
        <calcite-dropdown-item>Option 4</calcite-dropdown-item>
        </calcite-dropdown-group>
      </calcite-dropdown>
</calcite-button-with-dropdown>
```

This makes it a little less prescriptive (and OTOH perhaps open to more "error" by a consumer), but opens up the component to let folks use other `appearance` types of buttons and relies less on prop spreading. Of course this means the "divider" background color would need to be changed in some way. 

Overall this is a great addition, regarding `host-context` - we got word from Stencil they do not and do not plan to update their polyfill for that so we may need to reconfigure that and rollback the recent changes we made on Master :(

Also curious if other folks have opinions on naming of component?

🍿

kstinson14 commented:
@macandcheese I'd initially thought about doing a more slot-oriented approach like you suggested.  I decided on a more prop-oriented approach instead in order to 
1. keep the component less error-prone and ensure consistency a bit more (a `calcite-button` / `calcite-dropdown` were used as the slotted components, the `caretDown` icon is always used in the dropdown, etc).  I saw the more-prescriptive aspect as desirable, at least for now.  If the elements were slotted, users could use maybe an `outline` or `inline` appearance for the button as you say, which seems more customizable.  But the styling currently in place for this component is only meant for `solid` buttons, so while you _could_ use those variants on the buttons slotted into the component, the overall component wouldn't be well-styled, which could confuse the consumer if they thought they'd be able to use all permutations on a button here.  If we eventually wanted to style this to allow for all button appearances, we could reassess.  But I think for now, restricting what the user can pass in ensures they always end up with a viable component.
and 
2. Remove some of the boilerplate that users would have to write out when creating one of these components. If the button / dropdown were slotted, users would have to match their themes, colors, scale of buttons, scale of dropdown, alignment of dropdown, etc.  by themselves in their own code.  It wouldn't be difficult, but it's a bunch of extra stuff that the consumer has to write, and removes some of the benefits of having these bundled in one component to begin with, I feel.  This component especially doesn't have a lot of internal, functional code either.  Most of the code within it is meant to match the styling of the button and the dropdown, so I feel like getting rid of that might lose a lot of the component's value.

I do agree that we don't want to go crazy passing props into things, but I think in this case it's not an unreasonable amount, and we get some worthwhile styling cohesiveness out of it.  Let me know if this seems reasonable!

Oh shoot, I'll take out the `host-context` stuff then 👍 .

Definitely open to new names for the component!  I had trouble thinking of a good name for it.  I started with `calcite-button-with-overflow`, and a team member later suggested `calcite-button-with-dropdown` which seemed a bit more agnostic, so I went with that.

macandcheese commented on src/components/calcite-button-with-dropdown/calcite-button-with-dropdown.tsx:
```diff
@@ -0,0 +1,131 @@
+              color={this.color}
+              disabled={this.disabled}
+              theme={this.theme}
+              icon="caretDown"
```
Nitpick - can we allow folks to choose between chevron / caret down like we do in calcite-accordion: 

https://github.com/Esri/calcite-components/blob/1c59e44142d6d3456821f68729bb2712a1ef02eb/src/components/calcite-accordion/calcite-accordion.tsx#L47

kstinson14 replied:
👍 Makes sense to me.  Will change.

kstinson14 commented:
Made some changes:
- Added `dropdown-icon-type` property with `chevron` / `caret` choice to better mirror accordion component.  I left out the plus choice as it didn't seem to fit here as well.
- Removed `host-context` fixes since it seems we might be removing those selectors soon anyway.

paulcpederson commented on src/components/calcite-button-with-dropdown/readme.md:
```diff
@@ -0,0 +1,50 @@
+# calcite-button-with-dropdown
```
Can we add a description of what this component is for here? I'm not totally sure myself, so would be appreciated.

paulcpederson commented:
I will try to run the branch locally and check it out today

julio8a commented:
> Definitely open to new names for the component! I had trouble thinking of a good name for it.

Normally, this is called a split button.
https://www.nngroup.com/articles/split-buttons/

bstifle commented:
@kstinson14 hey nice work on these.

i would like to normalize all of the styling of the  |  divider on all the buttons to match what you did on the dark/light buttons

![image](https://user-images.githubusercontent.com/14083530/77564186-8afd9d00-6e7f-11ea-932e-512b003d58cb.png)

kstinson14 commented:
Thanks for the feedback guys!

@paulcpederson Will add a description 👍 

@julio8a I like that, `calcite-split-button` makes sense as a name to me.  

@bstifle  Oh shoot, thanks for the catch.  The padding on the dividers _should_ be the same.  A bug must've crept in somewhere in my recent changes.  Will change!  Will fix the weird colors on the light / dark controls as well.

kstinson14 commented:
@paulcpederson , @julio8a , @bstifle  Made some more changes:
- Added a readme description
- Renamed component to `calcite-split-button`
- Fixed divider sizes + background colors

driskull commented on src/components/calcite-split-button/calcite-split-button.tsx:
```diff
@@ -0,0 +1,145 @@
+  @Prop({ reflect: true }) disabled?: boolean;
+
+  /** fired when the modal begins the open animation */
+  @Event() primaryButtonClicked: EventEmitter;
```
Should the event name have `calciteSplitButton` in it? Like `calciteSplitButtonPrimaryClick`? I think it should be present tense rather than past tense as well.

kstinson14 replied:
Ah, that makes sense.  Would definitely make the event easier to track in the consuming app, will change.

driskull commented on src/components/calcite-split-button/calcite-split-button.tsx:
```diff
@@ -0,0 +1,145 @@
+    const dir = getElementDir(this.el);
+    return (
+      <Host dir={dir}>
+        <div>
```
Can we add a class onto this div so it doesn't look like it isn't necessary? Maybe a `.container` class that can be used in the`.scss` instead of `div`.

driskull replied:
`class="split-button__container"` maybe? :)

driskull commented on src/components/calcite-split-button/calcite-split-button.tsx:
```diff
@@ -0,0 +1,145 @@
+            dir={dir}
+            theme={this.theme}
+            scale={this.dropdownScale}
+            width={this.dropdownScale}
```
should the width be set to the scale here?

kstinson14 replied:
I thought the `'L'` dropdown scale + width was a bit too big for the current available sizes of this control.  

Originally I'd had `xs` and `xl` versions of the control (to match all available sizes of button), but the proportions between text and icon-only buttons didn't really work on the larger sizes.  So I removed the largest two (which used the `L` scale/width of dropdown), and made the smaller three `s`, `m`, and `l`.

If we were to ever add back these larger sizes, I think using the large dropdown there would make sense.

macandcheese replied:
The `width` attribute for dropdown is mutually exclusive from `scale` for dropdown, so those would be two separate pass-through-props I guess? For the scale... we will be updating the spacing on those a bit, but I feel like maybe it's a little weird that the pass-through-prop of `scale="l"` wouldn't result in a child dropdown with a `scale="l"`?

driskull commented on src/components/calcite-split-button/calcite-split-button.tsx:
```diff
@@ -0,0 +1,145 @@
+    return (
+      <Host dir={dir}>
+        <div>
+          <calcite-button
```
@kstinson14 What are your thoughts on having the `calcite-button` and `calcite-dropdown` as slots instead of being created internally and passing all the props? My concern is that as button adds props, this component needs to as well.

Maybe another option is to have just two props `buttonOptions` and `menuOptions` that could be typed as `CalciteButtonProperties`, etc.

Then it could just spread those options on the element like `<calite-button ...buttonOptions>`?

The slot might be more HTML consistent though but you would have to then query for the slotted element and set some of the props like iconPosition, onClick etc.

kstinson14 replied:
Ah, @macandcheese had a similar question which I answered above.  Basically, I see the main benefit of this component the stylistic combination of the two buttons and the dropdown.

The component lets the user pass in properties once, and have them propagate to all three sub-components.  Colors, scales, etc. are matched without making the user declare them all three times, and without the user needing to worry about inconsistencies between the three (scales for buttons go from `xs` to `xl`, but dropdowns only go `s` to `l` for example).

There's not much internal logic that the component provides, it's mainly this internal alignment of the component visuals.  If we were to slot in the button and dropdown (or use the options properties you describe) and make that alignment something the user had to do themselves, I feel like we'd be losing a lot of the value of the component.

I definitely agree that we should be cautious about how many properties we add to things.  But I don't think that's necessarily something we need to worry about here. I think the control that this component creates is fairly defined.  I don't think it _needs_ to support all of the properties/variations that `calcite-button` has or will receive.  There's several that it doesn't use now, and ones that it makes decisions about internally.  I don't think restricting the aspects of the buttons you can customize here is undesirable.

driskull replied:
> The component lets the user pass in properties once, and have them propagate to all three sub-components.

👍 

Agreed that this is worth not slotting then.

macandcheese replied:
Sounds good, we can always re-evaluate later if a need comes up for more button props to be used here.

macandcheese commented:
> I think we can get this in and make any necessary changes before we document/release it.

Agreed! Last thing would be naming - like @julio8a said it seems like most component libraries refer to this as `split-button` or `dropdown-button` / `button-dropdown`. None of which seem great either... but `split-button` might be the most common 🤷‍♂

julio8a commented:
Yes, let's go with split-button.

kstinson14 commented:
@driskull @macandcheese I think I've responded to all feedback:
- Renamed to `split-button`
- Renamed the button's event to include the component name
- Added class to container `div` to make it more apparent why it's there
- Left comments about dropdown size + prop reasoning

bstifle commented:
looks goooood @kstinson14  nice work
